src/number/n_unknown.jl

```diff
@@ -486,29 +507,89 @@ end
-   end
+# We wrap everything into an N_Field or N_Ring.
+# But if the ring or the element type is mutable, we have to wrap them first
+# into something mutable (RingWrapper and FieldWrapper)
+#
+# To keep it type stable, we use a @generated function to treat being mutable
+# as a property available at compile time.
 
-   if R isa Nemo.Field
-      if ok
-         return N_Field{T}(R)
-      else
-         RR = FieldWrapper{typeof(R), elem_type(R)}(R)
-         return N_Field{elem_type(RR)}(RR)
-      end
-   else
-      if ok
-         return N_Ring{T}(R)
-      else
-         RR = RingWrapper{typeof(R), elem_type(R)}(R)
-         return N_Ring{elem_type(RR)}(RR)
-      end
-   end
+if VERSION < v"1.7"
+  # This function was added in >= 1.7
+  ismutabletype(::Type{T}) where {T} = T.mutable
+end
+
+# - If R::T is mutable, the wrapper type should be T itself (no need for a
+#   wrapper)
+# - If R::T is immutable, the wrapper type is say Ring[Field]Wrapper
+#
+# Let S be the wrapper type. The wrapped object will be constructed
+# as S(R). This will work in the second case (because we have such
+# constructors), but not in the first case. There is in general no
+# (::Type{T})(R::T) = R constructor.
+# So we introduce a dummy type which just does
+# (::Type{_DummyWrapper{T}})(R::T) = R
+
+struct _DummyWrapper{T}
+end
+
+_DummyWrapper{T}(x::T) where {T} = x
+
+@generated function mutable_field_or_ring_type_wrapper(::Type{T}, ::Type{U}) where {T <: Nemo.Ring, U}
+  if ismutabletype(T) && ismutabletype(U)
+    return :(_DummyWrapper{T})
+  else 
+    if T <: Nemo.Field
+      S = FieldWrapper{T, U}
+    else
+      S = RingWrapper{T, U}
+    end
+    return :($S)
+  end
+end
+
+# The second wrapper
+function _second_wrapper(R::Nemo.Ring)
+  return N_Ring{elem_type(R)}(R)
+end
+
+function _second_wrapper(R::Nemo.Field)
+  return N_Field{elem_type(R)}(R)
+end
+
+# The second wrapper type (we need this to determine the type of the result)
+# To no do the @generated again, we just reuse the _DummyWrapper again.
+function _second_wrapper_type(::Type{S}) where {S <: Nemo.Field}
+  return N_Field{elem_type(S)}
+end
+
+function _second_wrapper_type(::Type{_DummyWrapper{S}}) where {S <: Nemo.Field}
+  return N_Field{elem_type(S)}
+end
+
+function _second_wrapper_type(::Type{S}) where {S <: Nemo.Ring}
+  return N_Ring{elem_type(S)}
+end
+
+function _second_wrapper_type(::Type{_DummyWrapper{S}}) where {S <: Nemo.Ring}
+  return N_Ring{elem_type(S)}
+end
+
+function CoefficientRing(R::U, cached::Bool = true) where {U <: Nemo.Ring}
+   wrappedtype = mutable_field_or_ring_type_wrapper(U, elem_type(U))
+   return AbstractAlgebra.get_cached!(CoefficientRingID, R, cached) do
+      # see comments for RingWrapper and FieldWrapper types
+      newring = wrappedtype(R) # this is either R or Ring[Field]Wrapper(R)
+      # Now comes the second layer of wrapping
+      RR = _second_wrapper(newring)
+      return RR
+    end::_second_wrapper_type(wrappedtype)
 end
```

Comment:
How about this instead? Seems to work OK in my (limited) tests.
```suggestion
@generated function mutable_field_or_ring_type_wrapper(::Type{T}, ::Type{U}) where {T <: Nemo.Ring, U}
  if ismutabletype(T) && ismutabletype(U)
    S = T
  elseif T <: Nemo.Field
    S = FieldWrapper{T, U}
  else
    S = RingWrapper{T, U}
  end
  return :($S)
end

function CoefficientRing(R::T, cached::Bool = true) where {T <: Nemo.Ring}
  U = elem_type(T)
  wrappertype = mutable_field_or_ring_type_wrapper(T, U)
  wrapperelemtype = elem_type(wrappertype)
  rettype = T <: Nemo.Field ? N_Field{wrapperelemtype} : N_Ring{wrapperelemtype}
  return AbstractAlgebra.get_cached!(CoefficientRingID, R, cached) do
    if ismutabletype(T) && ismutabletype(U)
      newring = R
    elseif T <: Nemo.Field
      newring = FieldWrapper{T, U}(R)
    else
      newring = RingWrapper{T, U}(R)
    end
    return rettype(newring)
  end::rettype
end
```

Reply:
Ah, you mean just making the anonymous function type unstable? Yes, that will also work if we do not care about that. We probably don't.

Reply:
Indeed, I think we don't care.  Overall, I see this:

<img width="898" alt="Screenshot 2022-01-13 at 12 20 31" src="https://user-images.githubusercontent.com/241512/149321108-3b0480c4-f2e1-4ffb-892a-268413810966.png">

The anonymous helper (made visible by a temporary hack to `CoefficientRing`):
<img width="897" alt="Screenshot 2022-01-13 at 12 23 44" src="https://user-images.githubusercontent.com/241512/149321578-ae2c27a6-72c7-40d3-b9e6-02db479c8742.png">